src/server.rs

```diff
@@ -101,9 +99,269 @@ pub enum Event {
+        FramebufferUpdate {
+            updates: Vec::new(),
+        }
+    }
+
+    /// Adds raw pixel data.
+    pub fn add_raw_pixels(&mut self, rect: Rect, pixel_data: &'a [u8]) -> &mut Self {
+        let update = Update::Raw {
+            rect: rect,
+            pixel_data: pixel_data
+        };
+
+        self.updates.push(update);
+        self
+    }
+
+    /// Adds `CopyRect` update message instructing client to reuse pixel data it already owns.
+    pub fn add_copy_rect(&mut self,
+                         dst: Rect,
+                         src_x_position: u16,
+                         src_y_position: u16)
+                         -> &mut Self {
+        let update = Update::CopyRect {
+            dst: dst,
+            src_x_position: src_x_position,
+            src_y_position: src_y_position,
+        };
+
+        self.updates.push(update);
+        self
+    }
+
+    /// Adds compressed pixel data.
+    ///
+    /// TODO: add method taking uncompressed data and compressing them.
+    pub fn add_compressed_pixels(&mut self, rect: Rect, zlib_data: &'a [u8]) -> &mut Self {
+        let update = Update::Zrle {
+            rect: rect,
+            zlib_data: zlib_data
+        };
+
+        self.updates.push(update);
+        self
+    }
+
+    /// Add data for drawing cursor.
+    pub fn add_cursor(&mut self,
+                      width: u16,
+                      height: u16,
+                      hotspot_x: u16,
+                      hotspot_y: u16,
+                      pixels: &'a [u8],
+                      mask_bits: &'a [u8])
+                      -> &mut Self {
+        let update = Update::SetCursor {
+            size: (width, height),
+            hotspot: (hotspot_x, hotspot_y),
+            pixels: pixels,
+            mask_bits: mask_bits
+        };
+
+        self.updates.push(update);
+        self
+    }
+
+    /// Adds notification about framebuffer resize.
+    pub fn add_desktop_size(&mut self, width: u16, height: u16) -> &mut Self {
+        let update = Update::DesktopSize {
+            width: width,
+            height: height,
+        };
+
+        self.updates.push(update);
+        self
+    }
+
+    /// Adds confirmation of support of pseudo-encoding.
+    pub fn add_pseudo_encoding(&mut self, encoding: protocol::Encoding) -> &mut Self {
+        let update = Update::Encoding { encoding: encoding };
+
+        self.updates.push(update);
+        self
+    }
+
+    /// Checks if all updates are valid.
+    ///
+    /// Panics if any of the updates is not valid.
+    fn check(&self, validation_data: &ValidationData) {
+        for update in self.updates.iter() {
+            update.check(validation_data);
+        }
+    }
+
+    /// Serializes this structure and sends it using given `writer`.
+    fn write_to<W: Write>(&self, writer: &mut W) -> Result<()> {
+        for chunk in self.updates.chunks(u16::max_value() as usize) {
+            let count = chunk.len() as u16;
```

Comment:
Why would I? Every chunk is now sent in separate message. There's no limit for number of messages. Alternative would be to limit self.update to u16::MAX updates in add_* methods.